ICAFIX/scripts/run_icasso.m

```diff
@@ -0,0 +1,253 @@
+S_final = S_final(:,Is);
+A_final = A_final(:,Is);
+iq = iq(Is);
+
+W_final = pinv(A_final);
+
+%% calculate single-regression z-stats
+NODEtsnorm = normalise(A_final);% z-score
+pN = pinv(NODEtsnorm); dpN = diag(pN * pN')';
+dof = size(NODEtsnorm,1) - size(NODEtsnorm,2) - 1;
+residuals = demean(vnts, 2) - S_final * NODEtsnorm';
+t = double(S_final ./ sqrt(sum(residuals .^ 2, 2) * dpN / dof));
+Z = zeros(size(t));
+Z(t > 0) = min(-norminv(tcdf(-t(t > 0), dof)),  38.5);
+Z(t < 0) = max( norminv(tcdf( t(t < 0), dof)), -38.5);
+Z(isnan(Z)) = 0;
+
+%% save icasso outputs to match melodic
+% reshape betas (S_final) and z-scores back to 4-d
+mtxDim = [mtxDim(1) Dim];
+volDim = [volDim(1:3) Dim];
+S_final4d = reshape2d4d(S_final,volDim,brainMask,mtxDim);
+Z4d = reshape2d4d(Z,volDim,brainMask,mtxDim);
+
+% save betas and z-score as 4-d nifti volumes
+hdr = infoNIFTI(vntsFile);
+hdr.ImageSize(end) = Dim;
+writeNIFTI(single(S_final4d),sprintf('%s/melodic_oIC',outDir),hdr)
+writeNIFTI(single(Z4d),sprintf('%s/melodic_IC',outDir),hdr)
+
+% save mixing matrices as tab-delimited text
+dlmwrite(sprintf('%s/melodic_unmix', outDir), W_final, '\t');
+dlmwrite(sprintf('%s/melodic_mix',   outDir), A_final, '\t');
+copyfile(sprintf('%s/melodic_mix',   outDir), sprintf('%s/melodic_Tmodes', outDir));% mix and Tmodes are the same
+
+% save variance explained as tab-delimited text
+ICstats = [tICAPercentVariances tICAPercentVariances * totVariance]; 
+dlmwrite([outDir '/melodic_ICstats'], ICstats, 'delimiter', '\t');
+
+% copy brainmask into melodic dir
+copyfile(brainMaskFile,[outDir '/mask.nii.gz'])
+
+% save pcaD and pcaE? Looks like FIX doesn't need them
+
+%% calculate and save FTmix spectra
+ts = struct();
+ts.Nsubjects = 1;
+ts.ts = A_final;
+[ts.NtimepointsPerSubject,ts.Nnodes] = size(ts.ts);
+ts_spectra = nets_spectra_sp(ts);
+dlmwrite([outDir '/melodic_FTmix'], ts_spectra, 'delimiter', '\t');
+fid = fopen([outDir '/components.txt'],'w');
+fprintf(fid,'%i: Signal\n',1:size(S_final,2));fclose(fid);
+
+% FTmix.sdseries.nii aren't need by FIX, so don't produce them
+% [~,~] = unix(['wb_command -cifti-create-scalar-series ' ...
+%   sprintf('%s/melodic_FTmix %s/melodic_FTmix.sdseries.nii -transpose -name-file %s/components.txt -series HERTZ 0 %f',...
+%   outDir,outDir,outDir,tr)]);
+
+%% Mixture modeling
+fprintf('performing melodic mixture modeling ...\n')
+mixtureModel([outDir '/melodic_IC.nii.gz']);% overwrites in-place without saving full report
+
+%% Helper subfunctions
+function [mtx,mtxDim] = reshape4d2d(img,msk)
+  % reshapes 4d img into 2d matrix where the first 3 dims are put into the 1st dim
+  % second arg msk is 3d logical volume
+  % second output is size before mask is applied
+
+  imgDim = size(img);
+  mtx = reshape(img,prod(imgDim(1:3)),imgDim(4));
+  mtxDim = size(mtx);%
+  if nargin == 2
+    mtx = mtx(logical(msk(:)),:);
+  end
+end
+
+function img = reshape2d4d(mtx,imgDim,msk,mtxDim)
```

Comment:
The "2d4d" versus "4d2d" naming difference is not very obvious.  Similar to my other naming suggestion, this could be called "unflattenData" instead.

Reply:
Renamed.